### PR TITLE
fix: replace tsutils with ts-api-tools

### LINKED ISSUE #5552: Repo: drop dependency on tsutils
### Suggestion

[tsutils](https://github.com/ajafff/tsutils) hasn't had a release (or even a commit) in well over a year (march 2021).

It seems like it's a dead package at this point, sadly - there are a number of issues going unanswered and PRs unreviewed. I guess that the owner has moved on to other things, and unfortunately they were the single point of failure.

This hasn't been a problem thus far, but with TS4.8 one of the types is now wrong and required a local patch to work around (#5551).
There are also some known issues caused by the package, which we cannot address without upstream patches (#5458, #3405).

I see three paths forward:
1) fork the package and release it under our namespace.
1) attempt to take ownership of the package and updated it.
1) bring the relevant utils internal into our `type-utils` package and drop the dependency.

(1) really conflicts with our `type-utils` package, so it seems immediately dead in the water.
(2) could work, but I personally don't have time to maintain it as a separate project - so we'd be relying on someone else for it (which doesn't really change our predicament - just shifts it). Also it would rely on the current owner being responsive or would rely on npm support passing ownership (which takes time).

IMO just dropping the dependency entirely and bringing the relevant utils into our `type-utils` package is the best course of action as it allows us to control the code that we so heavily rely on, which makes it easier for us to maintain everything. Given that we publish them as a package, the open-source community still benefits from our maintained versions.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

- [x] Addresses an existing open issue: fixes #5552
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Uses [`ts-api-tools`](https://github.com/JoshuaKGoldberg/ts-api-tools) as a drop-in replacement for [`tsutils`](https://github.com/ajafff/tsutils).

Note that `tsutils` provides some `is*` typeguards that are now exposed by TypeScript (e.g. `ts.isBinaryExpression`). I've gone ahead and switched to the TypeScript versions in this PR. Once #5573 is in we'll have CI checking to make sure our TypeScript version range supports them.

### CODE REVIEW

**Review by bradzacher — APPROVED** [review]
Can we trust this new package? I've heard some terrible things about the maintainer.

File: packages/eslint-plugin/src/rules/no-floating-promises.ts [hunk 1]
```diff
@@ -234,13 +234,13 @@ export default util.createRule<Options, MessageId>({
   },
 });
 
-// Modified from tsutils.isThenable() to only consider thenables which can be
+// Modified from ts-api-tools.isThenable() to only consider thenables which can be
 // rejected/caught via a second parameter. Original source (MIT licensed):
 //
-//   https://github.com/ajafff/tsutils/blob/49d0d31050b44b81e918eae4fbaf1dfe7b7286af/util/type.ts#L95-L125
+//   https://github.com/ajafff/ts-api-tools/blob/49d0d31050b44b81e918eae4fbaf1dfe7b7286af/util/type.ts#L95-L125
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/no-floating-promises.ts:240]: agh. there's always one!

File: jest.config.base.js [hunk 2]
```diff
@@ -11,6 +11,7 @@ module.exports = {
     'tsx',
     'mts',
     'mtsx',
+    'cjs',
```
Comment (JoshuaKGoldberg) [inline jest.config.base.js:14]: `ts-api-tools` formats its output files as `.js` with ES modules. But it also includes an additional set of `.cjs` files for CJS users. This allows Jest to consider `.cjs` files (which it didn't before) - and look at them _before_ `.js` files.

**Comment by JoshuaKGoldberg** [comment]
Note that I renamed the package to `ts-api-utils` in https://github.com/JoshuaKGoldberg/ts-api-utils/pull/28. We don't have to take action on the rename just yet. Just a note for the next time we need to bump the package version.

**Pull request merged**